File: 2024/Day6/2.py
from collections import deque
import itertools

def parse_input(grid):
    guard_pos = None
    guard_dir = None
    obstacles = set()
    directions = {'^': (0, -1), '>': (1, 0), 'v': (0, 1), '<': (-1, 0)}

    for y, line in enumerate(grid):
        for x, char in enumerate(line):
            if char == '#':
                obstacles.add((x, y))
            elif char in directions:
                guard_pos = (x, y)
                guard_dir = char

    return guard_pos, guard_dir, obstacles

def next_position(position, direction):
    x, y = position
    dx, dy = direction
    return x + dx, y + dy

def turn_right(direction):
    directions = ['^', '>', 'v', '<']
    idx = directions.index(direction)
    return directions[(idx + 1) % 4]

def simulate_path(start_pos, start_dir, obstacles, max_steps=10000):
    directions = {'^': (0, -1), '>': (1, 0), 'v': (0, 1), '<': (-1, 0)}

    pos = start_pos
    dir_char = start_dir
    visited = set()
    steps = 0

    while steps < max_steps:
        if (pos, dir_char) in visited:
            return True  # Loop detected
        visited.add((pos, dir_char))
        next_pos = next_position(pos, directions[dir_char])

        if next_pos in obstacles:
            # Turn right
            dir_char = turn_right(dir_char)
        else:
            # Move forward
            pos = next_pos

        steps += 1

    return False

def find_loop_positions(grid):
    guard_pos, guard_dir, obstacles = parse_input(grid)
    possible_positions = set()

    # Precompute all empty positions
    empty_positions = [(x, y) for y, line in enumerate(grid) for x, char in enumerate(line) if char == '.' and (x, y) != guard_pos]

    for x, y in empty_positions:
        # Try placing an obstruction at (x, y)
        new_obstacles = obstacles | {(x, y)}
        if simulate_path(guard_pos, guard_dir, new_obstacles):
            possible_positions.add((x, y))

    return len(possible_positions)

# Read input from file
with open('input.txt', 'r') as f:
    grid = [line.strip() for line in f.readlines()]

result = find_loop_positions(grid)
print("Number of possible positions to create a loop:", result)

File: 2024/Day6/2.swift
import Foundation

struct Position: Hashable {
    let x: Int
    let y: Int
}

struct GuardSimulation {
    var guardPos: Position
    var guardDir: Character
    var obstacles: Set<Position>
    let directions: [Character: (x: Int, y: Int)] = [
        "^": (0, -1), ">": (1, 0), "v": (0, 1), "<": (-1, 0)
    ]

    init(grid: [String]) {
        var guardPosition: Position? = nil
        var guardDirection: Character? = nil
        var obstaclesSet = Set<Position>()

        for (y, line) in grid.enumerated() {
            for (x, char) in line.enumerated() {
                if char == "#" {
                    obstaclesSet.insert(Position(x: x, y: y))
                } else if directions.keys.contains(char) {
                    guardPosition = Position(x: x, y: y)
                    guardDirection = char
                }
            }
        }

        self.guardPos = guardPosition!
        self.guardDir = guardDirection!
        self.obstacles = obstaclesSet
    }

    func nextPosition(_ position: Position, _ direction: (x: Int, y: Int)) -> Position {
        return Position(x: position.x + direction.x, y: position.y + direction.y)
    }

    func turnRight(_ direction: Character) -> Character {
        let directionsOrder: [Character] = ["^", ">", "v", "<"]
        if let idx = directionsOrder.firstIndex(of: direction) {
            return directionsOrder[(idx + 1) % directionsOrder.count]
        }
        return direction
    }

    func simulatePath(startPos: Position, startDir: Character, obstacles: Set<Position>, maxSteps: Int = 10000) -> Bool {
        var pos = startPos
        var dirChar = startDir
        var visited = Set<VisitedState>()
        var steps = 0

        while steps < maxSteps {
            let currentState = VisitedState(position: pos, direction: dirChar)
            if visited.contains(currentState) {
                return true // Loop detected
            }
            visited.insert(currentState)
            let direction = directions[dirChar]!
            let nextPos = nextPosition(pos, direction)

            if obstacles.contains(nextPos) {
                // Turn right
                dirChar = turnRight(dirChar)
            } else {
                // Move forward
                pos = nextPos
            }

            steps += 1
        }

        return false
    }

    func findLoopPositions(grid: [String]) -> Int {
        var possiblePositions = Set<Position>()

        // Precompute all empty positions
        var emptyPositions = [Position]()
        for (y, line) in grid.enumerated() {
            for (x, char) in line.enumerated() {
                if char == "." && Position(x: x, y: y) != guardPos {
                    emptyPositions.append(Position(x: x, y: y))
                }
            }
        }

        let checkPosition: (Position) -> Position? = { position in
            var newObstacles = self.obstacles
            newObstacles.insert(position)
            return self.simulatePath(startPos: self.guardPos, startDir: self.guardDir, obstacles: newObstacles) ? position : nil
        }

        let results = emptyPositions.compactMap { checkPosition($0) }
        possiblePositions = Set(results)

        return possiblePositions.count
    }
}

struct VisitedState: Hashable {
    let position: Position
    let direction: Character
}

let fileURL = URL(fileURLWithPath: "input.txt")
// Read input from file
if let content = try? String(contentsOf: fileURL, encoding: .utf8) {
    let grid = content.split(separator: "\n").map { String($0) }
    let simulation = GuardSimulation(grid: grid)
    let result = simulation.findLoopPositions(grid: grid)
    print("Number of possible positions to create a loop: \(result)")
}
